src/persist-client/src/internal/gc.rs

```diff
@@ -340,21 +342,19 @@ where
         // becomes an issue. Maybe make Blob::delete take a list of keys?
         //
         // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
-        //
-        // Another idea is to use a FuturesUnordered to at least run them
-        // concurrently, but this requires a bunch of Arc cloning, so wait to
-        // see if it's worth it.
+        let mut futures = vec![];
+        let blob: &(dyn Blob + Send + Sync) = machine.state_versions.blob.borrow();
         for key in deleteable_batch_blobs {
-            retry_external(&machine.metrics.retries.external.batch_delete, || async {
-                machine
-                    .state_versions
-                    .blob
-                    .delete(&key.complete(&req.shard_id))
-                    .await
-            })
-            .instrument(debug_span!("batch::delete"))
-            .await;
+            let key = key.complete(&req.shard_id);
+            futures.push(
+                retry_external(&machine.metrics.retries.external.batch_delete, move || {
+                    let key = key.clone();
+                    async move { blob.delete(&key).await }
+                })
+                .instrument(debug_span!("batch::delete")),
+            )
         }
+        join_all(futures).await;
```

Comment:
I'd feel more comfortable if we had a limit on the concurrency of these requests. Steady-state we'll be fine, but if we're chewing through a large backlog (e.g. seqno hold bug, or a long-running snapshot finishes) I wouldn't want us to start hitting 503s for more sensitive requests due to DoSing our S3 client with deletes. Maybe we can toss in a tokio Semaphore with a fixed # of permits?

Reply:
fwiw, I considered making the same comment and realized I think I'm happy either way. we don't have concurrency limits in other places yet, so it seems somewhat arbitrary to me to start here. (I'm also a touch skeptical of local concurrency limits vs process-wide ones but the latter is certainly a dramatic scope increase for this PR)

Reply:
Agreed that fully auditing our API calls to be conscientious of rate limits is a much larger task, but IMO this one is straightforward to reason about, e.g. would we want to explain to a customer that we temporarily couldn't read/write live data because we wanted to delete unused data as quickly as possible? (probably not) There are some places that already effectively have concurrency limits too, like how batch building uses `batch_builder_max_outstanding_parts` to limit the # of inflight reqs we have to blob storage.

Local vs process-wide rate limits is a good point, and one could even zoom out further to environment-wide rate limits when thinking about S3 prefix limits. Let's table that one for later 😛

Reply:
wfm!

Reply:
I've wrapped this up in a semaphore, and also switched to `FuturesUnordered`... supposedly somewhat more efficient when dealing with large numbers of futures, which if we're worried about knocking over S3 is presumably the case!

(Really ought to live in the Blob impl itself, but that can come later.)